priiloader/source/Input.cpp

```diff
@@ -72,19 +129,39 @@ s8 Input_Init( void )
 	gettimeofday(&_time_init, NULL);
 
 	STM_RegisterEventHandler(HandleSTMEvent);
+	r |= KEYBOARD_Init(HandleKBDEvent);
+	// Even if the thread fails to start, we shouldn't call KEYBOARD_Deinit because
+	// it's broken and calling it twice will crash and it's already called in Input_Shutdown
+	kbd_should_quit = false;
+	LWP_CreateThread(	&kbd_handle,
+							kbd_thread,
+							NULL,
+							NULL,
+							16*1024,
+							50);
 
 	_input_init = 1;
 	return r;
 }
 void Input_Shutdown( void )
 {
+	if (_input_init != 1)
+		return;
+	
 	WPAD_Shutdown();
+	kbd_should_quit = true;
+	if (kbd_handle != LWP_THREAD_NULL) {
+		LWP_JoinThread(kbd_handle, NULL);
+		kbd_handle = LWP_THREAD_NULL;
+	}
+	KEYBOARD_Deinit();
 	_input_init = 0;
 }
 u32 Input_ScanPads( void )
 {
 	WPAD_ScanPads();
 	PAD_ScanPads();
+	usleep(1000); // Give the keyboard thread a chance to run
```

Comment:
a 1s sleep sounds very dangerous and sounds it'll fail if youre fast pressing a button on a wiimote, only doing something every second

Reply:
This is something I would like to fix, unfortunately I'm not sure of another way, LWP_Yield doesn't give it enough time either. Also this isn't 1s, usleep is microseconds. I haven't experienced any input drop because of this.